internal/cmd/update.go

```diff
@@ -0,0 +1,466 @@
+package cmd
+
+import (
+	"archive/tar"
+	"archive/zip"
+	"bufio"
+	"bytes"
+	"compress/gzip"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"time"
+
+	"cdr.dev/coder-cli/internal/config"
+	"cdr.dev/coder-cli/internal/version"
+	"cdr.dev/coder-cli/pkg/clog"
+	"golang.org/x/xerrors"
+
+	"github.com/Masterminds/semver/v3"
+	"github.com/manifoldco/promptui"
+	"github.com/spf13/afero"
+	"github.com/spf13/cobra"
+)
+
+const (
+	goosWindows       = "windows"
+	goosLinux         = "linux"
+	apiPrivateVersion = "/api/private/version"
+)
+
+// updater updates coder-cli.
+type updater struct {
+	confirmF       func(string) (string, error)
+	execF          func(context.Context, string, ...string) ([]byte, error)
+	executablePath string
+	fs             afero.Fs
+	httpClient     getter
+	osF            func() string
+	versionF       func() string
+}
+
+func updateCmd() *cobra.Command {
+	var (
+		force      bool
+		coderURL   string
+		versionArg string
+	)
+
+	cmd := &cobra.Command{
+		Use:   "update",
+		Short: "Update coder binary",
+		Long:  "Update coder to the version matching a given coder instance.",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			ctx := cmd.Context()
+			httpClient := &http.Client{
+				Timeout: 10 * time.Second,
+			}
+
+			currExe, err := os.Executable()
+			if err != nil {
+				return clog.Fatal("init: get current executable", clog.Causef(err.Error()))
+			}
+
+			updater := &updater{
+				confirmF:       defaultConfirm,
+				execF:          defaultExec,
+				executablePath: currExe,
+				httpClient:     httpClient,
+				fs:             afero.NewOsFs(),
+				osF:            func() string { return runtime.GOOS },
+				versionF:       func() string { return version.Version },
+			}
+			return updater.Run(ctx, force, coderURL, versionArg)
+		},
+	}
+
+	cmd.Flags().BoolVar(&force, "force", false, "do not prompt for confirmation")
+	cmd.Flags().StringVar(&coderURL, "coder", "", "query this coder instance for the matching version")
+	cmd.Flags().StringVar(&versionArg, "version", "", "explicitly specify which version to fetch and install")
+
+	return cmd
+}
+
+type getter interface {
+	Get(url string) (*http.Response, error)
+}
+
+func (u *updater) Run(ctx context.Context, force bool, coderURLArg string, versionArg string) error {
+	// Check under following directories and warn if coder binary is under them:
+	//   * C:\Windows\
+	//   * homebrew prefix
+	//   * coder assets root (/var/tmp/coder)
+	var pathBlockList = []string{
+		`C:\Windows\`,
+		`/var/tmp/coder`,
+	}
+	brewPrefixCmd, err := u.execF(ctx, "brew", "--prefix")
+	if err == nil { // ignore errors if homebrew not installed
+		pathBlockList = append(pathBlockList, strings.TrimSpace(string(brewPrefixCmd)))
+	}
+
+	for _, prefix := range pathBlockList {
+		if HasFilePathPrefix(u.executablePath, prefix) {
```

Comment:
This behavior is good for now, just wanted to leave a suggestion for your consideration in a future iteration: we could also write the file to a well-known location (e.g. `~/.local/coder/bin`) or similar, and then use LookPath to see if we find the file we just wrote. If the version doesn't match, then we know that the user has something else earlier on their path obscuring the location of where we wrote the file, so we could issue a message in that case instructing users what to do (e.g. move ~/.local/coder/bin earlier in the PATH, or remove the other one from the PATH, etc)